### PR TITLE
fix(skyrim-platform): execute game functions in game thread

### PR DESCRIPTION
**Pull request by Wedmer**
Should help with skyrim-multiplayer/skymp#653

### CODE REVIEW

File: skyrim-platform/src/platform_se/skyrim_platform/SkyrimPlatform.h [hunk 1]
```diff
@@ -13,6 +16,15 @@ class SkyrimPlatform
   void AddTickTask(const std::function<void()>& f);
   void AddUpdateTask(const std::function<void()>& f);
   void PushAndWait(const std::function<void(int)>& task);
+  void Push(const std::function<void(int)>& task);
+  void Pull(RE::BSTSmartPointer<RE::BSScript::IFunction> fPtr,
```
Comment (Pospelove) [inline skyrim-platform/src/platform_se/skyrim_platform/SkyrimPlatform.h:20]: `Pull` and `Push` feel like antonyms here but they are not. It's confusing.
Please rename `Pull` to `ExecuteInGameThread` or something like this, more insightful for code readers.

**Review by Pospelove — CHANGES_REQUESTED** [review]
The solution overall looks good and pretty clean. 👍

File: skyrim-platform/src/platform_se/skyrim_platform/CallNative.cpp [hunk 2]
```diff
@@ -17,6 +17,7 @@
 #include <skse64/GameReferences.h>
 #include <skse64/PapyrusActor.h>
 #include <skse64_common/Relocation.h>
+#include "SkyrimPlatform.h"
```
[the remark this answers is not in the document]
Comment (Wedmer) [reply skyrim-platform/src/platform_se/skyrim_platform/CallNative.cpp]: Done

File: skyrim-platform/src/platform_se/skyrim_platform/SkyrimPlatform.cpp [hunk 3]
```diff
@@ -302,12 +305,30 @@ class CommonExecutionListener : public TickListener
 };
 }
 
+typedef asio::executor_work_guard<asio::io_context::executor_type> spWork;
```
[the remark this answers is not in the document]
Comment (Wedmer) [reply skyrim-platform/src/platform_se/skyrim_platform/SkyrimPlatform.cpp]: Renamed

File: skyrim-platform/src/platform_se/skyrim_platform/SkyrimPlatform.cpp [hunk 4]
```diff
@@ -302,12 +305,30 @@ class CommonExecutionListener : public TickListener
...
+                RE::BSScript::ErrorLogger* a_logger,
+                RE::BSScript::Internal::VirtualMachine* vm,
+                RE::BSScript::IFunction::CallResult* ret)
+  {
+    *ret = fPtr->Call(a_stack, a_logger, vm, false);
+    std::lock_guard<std::mutex> lock(spLock);
+    complete = true;
+    spCV.notify_all();
+  }
+  bool complete;
```
[the remark this answers is not in the document]
Comment (Wedmer) [reply skyrim-platform/src/platform_se/skyrim_platform/SkyrimPlatform.cpp]: Initializes in SkyrimPlatform::SkyrimPlatform()

File: skyrim-platform/src/platform_se/skyrim_platform/SkyrimPlatform.cpp [hunk 5]
```diff
@@ -302,12 +305,30 @@ class CommonExecutionListener : public TickListener
...
 struct SkyrimPlatform::Impl
 {
   std::shared_ptr<BrowserApi::State> browserApiState;
   std::vector<std::shared_ptr<TickListener>> tickListeners;
   Viet::TaskQueue tickTasks, updateTasks;
   ThreadPoolWrapper pool;
+  asio::io_context spContext;
+  std::mutex spLock;
+  std::condition_variable spCV;
+  void spWorker(RE::BSTSmartPointer<RE::BSScript::IFunction> fPtr,
```
[the remark this answers is not in the document]
Comment (Wedmer) [reply skyrim-platform/src/platform_se/skyrim_platform/SkyrimPlatform.cpp]: Renamed

File: skyrim-platform/src/platform_se/skyrim_platform/SkyrimPlatform.cpp [hunk 6]
```diff
@@ -302,12 +305,30 @@ class CommonExecutionListener : public TickListener
...
 
+typedef asio::executor_work_guard<asio::io_context::executor_type> spWork;
+
 struct SkyrimPlatform::Impl
 {
   std::shared_ptr<BrowserApi::State> browserApiState;
   std::vector<std::shared_ptr<TickListener>> tickListeners;
   Viet::TaskQueue tickTasks, updateTasks;
   ThreadPoolWrapper pool;
+  asio::io_context spContext;
```
[the remark this answers is not in the document]
Comment (Wedmer) [reply skyrim-platform/src/platform_se/skyrim_platform/SkyrimPlatform.cpp]: Renamed

File: skyrim-platform/src/platform_se/skyrim_platform/SkyrimPlatform.cpp [hunk 7]
```diff
@@ -302,12 +305,30 @@ class CommonExecutionListener : public TickListener
...
+  std::mutex spLock;
+  std::condition_variable spCV;
+  void spWorker(RE::BSTSmartPointer<RE::BSScript::IFunction> fPtr,
+                const RE::BSTSmartPointer<RE::BSScript::Stack>& a_stack,
```
[the remark this answers is not in the document]
Comment (Wedmer) [reply skyrim-platform/src/platform_se/skyrim_platform/SkyrimPlatform.cpp]: Renamed

File: skyrim-platform/src/platform_se/skyrim_platform/SkyrimPlatform.cpp [hunk 8]
```diff
@@ -359,3 +380,41 @@ void SkyrimPlatform::PushAndWait(const std::function<void(int)>& f)
...
+}
+
+void SkyrimPlatform::StopWorker()
+{
+  pImpl->spContext.stop();
+}
+
+void SkyrimPlatform::PrepareWorker()
+{
+  if (pImpl->spContext.stopped())
```
[the remark this answers is not in the document]
Comment (Wedmer) [reply skyrim-platform/src/platform_se/skyrim_platform/SkyrimPlatform.cpp]: Done

File: skyrim-platform/src/platform_se/skyrim_platform/SkyrimPlatform.cpp [hunk 9]
```diff
@@ -359,3 +380,41 @@ void SkyrimPlatform::PushAndWait(const std::function<void(int)>& f)
...
+    pImpl->spContext.restart();
+  pImpl->wg.reset(new spWork(pImpl->spContext.get_executor()));
```
[the remark this answers is not in the document]
Comment (Wedmer) [reply skyrim-platform/src/platform_se/skyrim_platform/SkyrimPlatform.cpp]: Changed

File: skyrim-platform/src/platform_se/skyrim_platform/SkyrimPlatform.h (hunk 1 above)
Comment (Wedmer) [reply skyrim-platform/src/platform_se/skyrim_platform/SkyrimPlatform.h]: Renamed

**Review by Pospelove — APPROVED** [review]

**Pull request merged**